graphql_compiler/compiler/helpers.py

```diff
@@ -140,36 +140,51 @@ def get_uniquely_named_objects_by_name(object_list):
 
 def safe_quoted_string(value):
     """Return the provided string, surrounded by single quotes. Unsafe strings cause exceptions."""
-    validate_safe_string(value)
+    validate_safe_or_special_string(value)
     return u'\'{}\''.format(value)
 
 
-def validate_safe_string(value):
-    """Ensure the provided string does not have illegal characters."""
+def validate_safe_or_special_string(value, value_description='string'):
+    """Ensure the string does not have illegal characters or is in a set of allowed strings."""
     # The following strings are explicitly allowed, despite having otherwise-illegal chars.
     legal_strings_with_special_chars = frozenset({'@rid', '@class', '@this', '%'})
+    if value not in legal_strings_with_special_chars:
+        validate_safe_string(value, value_description='string')
 
-    if not isinstance(value, six.string_types):
-        raise TypeError(u'Expected string value, got: {} {}'.format(
-            type(value).__name__, value))
 
+def validate_safe_string(value, value_description='string'):
+    """Ensure that the provided string not have illegal characters."""
     if not value:
-        raise GraphQLCompilationError(u'Empty strings are not allowed!')
+        raise GraphQLCompilationError(u'Empty {}s are not allowed!'.format(value_description))
 
     if value[0] in string.digits:
-        raise GraphQLCompilationError(u'String values cannot start with a digit: {}'.format(value))
+        raise GraphQLCompilationError(u'Encountered invalid {}: {}. It cannot start with a '
+                                      u'digit.'.format(value_description, value))
+
+    if not set(value).issubset(VARIABLE_ALLOWED_CHARS):
+        raise GraphQLCompilationError(u'Encountered illegal characters in {}: {}. It is only '
+                                      u'allowed to have upper and lower case letters, '
+                                      u'digits and underscores. '
```

Comment:
Trailing space in the message. Also, let's refactor this to say which specific disallowed characters were present:
```
disallowed_chars = frozenset(value) - VARIABLE_ALLOWED_CHARS
if disallowed_chars:
    ...
```